### PR TITLE
HPCC-23202 Update Kafka plugin to use latest released Kafka client library

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Manual testing against latest Kafka server and client libraries.

### CODE REVIEW

**Comment by dcamper** [comment]
@ghalliday Please review or reassign.  Thanks!

File: plugins/kafka/kafka.cpp [hunk 1]
```diff
@@ -506,8 +526,17 @@ namespace KafkaPlugin
             consumerPtr->stop(topicPtr, partitionNum);
         }
 
-        delete(topicPtr.load());
-        delete(consumerPtr);
+        if (topicPtr)
```
Comment (ghalliday) [inline plugins/kafka/kafka.cpp]: Why this change?  You don't need to protect delete of a null pointer, and this is within the destructor so the values will never be accessed.

File: plugins/kafka/kafka.cpp [hunk 2]
```diff
@@ -506,8 +526,17 @@ namespace KafkaPlugin
...
+        {
+            delete(topicPtr.load());
+            topicPtr = NULL;
+        }
+
+        if (consumerPtr)
```
Comment (ghalliday) [inline plugins/kafka/kafka.cpp]: same here

File: plugins/kafka/kafka.cpp [hunk 3]
```diff
@@ -591,45 +624,41 @@ namespace KafkaPlugin
...
             }
         }
-        else
+        catch (...)
         {
-            throw MakeStringException(-1, "Kafka: Failed to start Consumer read for %s:%d @ %s; error: %d", topic.c_str(), partitionNum, brokers.c_str(), startErr);
+            // The current object is allocated on the heap and was supposed to be managed by the
+            // IRowStream subclass we're returning; since we've encounted an error, we need
+            // to self-destruct to avoid memory leaks
+            delete(this);
```
Comment (ghalliday) [inline plugins/kafka/kafka.cpp]: This rings some alarm bells.  Better would be to have the catch in the global getMessageDataset() function and delete consumerObjPtr.  (or use std::unique_ptr)

**Comment by ghalliday** [comment]
@dcamper a few comments/questions

File: plugins/kafka/kafka.cpp (hunk 1 above)
Comment (dcamper) [reply plugins/kafka/kafka.cpp]: Will revert.

File: plugins/kafka/kafka.cpp (hunk 2 above)
Comment (dcamper) [reply plugins/kafka/kafka.cpp]: Will revert.

File: plugins/kafka/kafka.cpp (hunk 3 above)
Comment (dcamper) [reply plugins/kafka/kafka.cpp]: Moving to using a catch in the calling function.

**Comment by dcamper** [comment]
Changes pushed.  @ghalliday Please recheck.  Thanks!

**Comment by ghalliday** [comment]
looks good to me.  @richardkchapman letting you double check before merging.
